src/commands/tick-stop.js

```diff
@@ -16,53 +17,102 @@ function builder(yargs) {
+  .then(td => {
+    timersDoc = td
+    return timersDoc.timers.pop()
+  })
+  .then(timer => ensureEndAndMessage(timer, message))
+  .then(timer => db.put(timersDoc).then(() => timer)) // Persist the timer pop
   .then(timer => commitTimer(db, timer))
   .then(writeSaved)
   .catch(err => console.error(`Could not stop your timer\n${err.message}`))
 }
 
-function parseMessage(timersDoc, newMessage) {
-  const timer = timersDoc.timers.pop()
+/*
+ * Recursive function to ensure user provides both an end time and message
+ *
+ * This is complicated as the user can input in several different ways and
+ * orders. As well as being able to cancel at any time which should result in
+ * the timer remaining in the couch document.
+ *
+ * Since we don't have tests for our commands, I will try to list the different
+ * cases:
+ *
+ * 1. Timer already has a message.
+ *   a. User provides an end time with the stop command
+ *      * Parse the end time from the new message
+ *   b. User provides a new commit message with no time
+ *      * Use current time for end
+ *      * Update message
+ *   c. User provides end time and message with stop command
+ *      * Use provided end time
+ *      * Update message
+ *   d. User does not provide a message with the stop command
+ *      * Use current time for end
+ * 2. Timer does not have a message
+ *   a. User provides an end time with the stop command
+ *      aa. Prompt user and they provide a message
+ *          * Set message
+ *          * Use originally stated end time
+ *      ab. Prompt user and they provide a message and new end time
+ *          * Set message
+ *          * Use new end time
+ *      ac. Prompt user and they only provide a new end time
+ *          * Update end time
+ *          * Prompt again for message
+ *   b. User provides message but no end time
+ *      * Set message
+ *      * User current time
+ *   c. User does not provide a message with the stop command
+ *      ca. Prompt user and they provide a message
+ *          * Set message
+ *          * Use current time for end
+ *      cb. Prompt user and they provide an end time
+ *          * Update end time
+ *          * Prompt again for message
+ *      cc. Prompt user and they provide end time and message
+ *          * Set end time
+ *          * Set message
+ */
+function ensureEndAndMessage(timer, newMessage) {
+  if (!newMessage && !timer.message) {
+    return promptForMessage()
+    .then(message => ensureEndAndMessage(timer, message))
+  } else if (newMessage) {
+    const { start: end, message } = parser(newMessage)
 
-  if (!timer) throw new Error('You do not have a timer started')
+    timer.end = end || timer.end || new Date()
+
+    if (!message && !timer.message) {
+      return promptForMessage()
+      .then(m => ensureEndAndMessage(timer, m))
+    } else {
+      timer.message = message || timer.message
+      return timer
+    }
+  } else {
+    timer.end = new Date()
+
+    return timer
+  }
```

Comment:
My thought with this would be that it's nice to have the default `else`. To do this without `else if` I would just not wrap the last block in an `if` statement. The trouble is that the returns above are async, so I thought it wouldn't return early until the async call is done. Is that a misunderstanding on my part? Will it always return early even if it calls an async function?